### PR TITLE
handle unhandled errors

### PR DESCRIPTION
**Pull request by shibumi**
*Please fill in the fields below to submit a pull request.  The more
information that is provided, the better.*

**Fixes issue #**:

This is not yet related to an issue. Shall I open an issue for it?

**Description of pull request**:

This PR fixes a few golang linter warnings like unhandled errors.
It's work in progress.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

New tests shouldn't be necessary.

### CODE REVIEW

**Comment by shibumi** [comment]
Mhh shouldn't there be a 'make this PR a draft' button somewhere? Or is this beta feature not enabled in toto-golang yet?

**Comment by shibumi** [comment]
Small note: I will squash the commits in the end, that way we will end up with a most atomic PR as possible and keep our git history clean.

**Comment by lukpueh** [comment]
Thanks, @shibumi!

> This is not yet related to an issue. Shall I open an issue for it?

I don't think that's necessary.

> squash the commits in the end

💯 for commit discipline! If your PR has a range of meaningful commits, there is no need to squash them.

**Comment by shibumi** [comment]
@lukpueh Thanks for your comment, ok Then I will continue fixing this up. I also have the feeling that our tests can't be parallelized (not sure if this is an issue, but we should at least *know* about this, and maybe fix this in the future).

**Comment by shibumi** [comment]
Ok, this commits should fix up nearly all errors except:

* error format violations.. erorrs should start with lowercase and should has no period. However if I change these messages, the tests start failing (I guess we are matching strings somewhere..)

* Code like this:
```go
	if err.Error() != "invalid type for link 'test_type': should be 'link'" {
		t.Error("validateLink error - incorrect type not detected")
	}
```

The call of "err.Error()" may lead to nil pointer dereference in very rare cases.. That's what the warning is telling me, but I have no idea how to fix this and if we even need to fix this.

**Comment by shibumi** [comment]
I think we can merge this for now, I am just not sure about the decreasing coverage.

**Comment by lukpueh** [comment]
> I also have the feeling that our tests can't be parallelized

Are you sure? Off the top of my head, I would say that it should be possible. But I'd need to double check.  If it's a problem, feel free to submit an issue.

> I also hope it's fine that I sign my commits

Of course it is. :)

> I am just not sure about the decreasing coverage.

You can see which lines aren't covered on [coveralls](https://coveralls.io/builds/31139615), or you [check coverage locally](https://blog.golang.org/cover). If it's feasible to add a test, please do so.

> squash the commits in the end

May I pick you up on that offer after all? :) You could for instance squash the multiple "handle unhandled error" commits into one.

**Comment by lukpueh** [comment]
> The call of "err.Error()" may lead to nil pointer dereference in very rare cases.

In the case you pasted you could just do,`if err == nil || err != "invalid ...`. Since the test expects a particular error message, it definitely must also fail if there's no error at all.

**Comment by lukpueh** [comment]
Btw. would you mind enabling the linter you use in .travis.yml, so that we don't merge in code with warnings?

**Comment by shibumi** [comment]
Hi @lukpueh thanks for your review

> Are you sure? Off the top of my head, I would say that it should be possible. But I'd need to double check. If it's a problem, feel free to submit an issue.

I should be able to test this with `go test -parallel -v ./...` right? I will check this and add a comment afterwards.

> You can see which lines aren't covered on coveralls, or you check coverage locally. If it's feasible to add a test, please do so.

Oh, I will have a look with go cover. I didn't know that I can check this locally with go. Thanks for your tip :+1: 

> May I pick you up on that offer after all? :) You could for instance squash the multiple "handle unhandled error" commits into one.

Sure, I that's why I offered it :) my commits are a little bit chaotic sometimes, so I definitely need to squash them. I just don't know if I can squash commits together that are not not directly neighbors. I need to check this with git.

> In the case you pasted you could just do,if err == nil || err != "invalid .... Since the test expects a particular error message, it definitely must also fail if there's no error at all.

Oh I will have a look.

> Btw. would you mind enabling the linter you use in .travis.yml, so that we don't merge in code with warnings?

It's the Jetbrains GoLand IDE linter.. no idea if this linter is open source, but I guess it's just using https://github.com/golang/lint I will verify this and ask around if this is true.

**Comment by shibumi** [comment]
@lukpueh Is the new git history okay or shall I squash more commits? Will have a look on the coverage next, although I can't explain why the commits have a bad effect on the coverage oO I have just added a few more error messages.

**Comment by shibumi** [comment]
oh and btw `golint` found a big bunch of other things my IDE didn't show me:
```
in_toto/canonicaljson.go:1:1: don't use an underscore in package name
in_toto/canonicaljson_test.go:1:1: don't use an underscore in package name
in_toto/examples_test.go:1:1: don't use an underscore in package name
in_toto/hash.go:1:1: don't use an underscore in package name
in_toto/hash.go:21:6: exported type Hash should have comment or be unexported
in_toto/hash_test.go:1:1: don't use an underscore in package name
in_toto/in_toto_test.go:1:1: don't use an underscore in package name
in_toto/keylib.go:1:1: don't use an underscore in package name
in_toto/keylib.go:93:2: var keyIdHashAlgorithms should be keyIDHashAlgorithms
in_toto/keylib_test.go:1:1: don't use an underscore in package name
in_toto/model.go:1:1: don't use an underscore in package name
in_toto/model.go:43:2: struct field KeyId should be KeyID
in_toto/model.go:44:2: struct field KeyIdHashAlgorithms should be KeyIDHashAlgorithms
in_toto/model.go:90:2: struct field KeyId should be KeyID
in_toto/model.go:194:7: exported const LinkNameFormatShort should have comment or be unexported
in_toto/model.go:195:7: exported const SublayoutLinkDirFormat should have comment or be unexported
in_toto/model.go:305:9: range var keyId should be keyID
in_toto/model.go:338:1: comment on exported method Layout.StepsAsInterfaceSlice should be of the form "StepsAsInterfaceSlice ..."
in_toto/model.go:350:1: exported method Layout.InspectAsInterfaceSlice should have comment or be unexported
in_toto/model.go:372:6: range var keyId should be keyID
in_toto/model.go:385:10: if block ends with a return statement, so drop this else and outdent its block
in_toto/model.go:395:10: if block ends with a return statement, so drop this else and outdent its block
in_toto/model.go:427:6: func checkRequiredJsonFields should be checkRequiredJSONFields
in_toto/model_test.go:1:1: don't use an underscore in package name
in_toto/model_test.go:22:2: var invalidJsonBytes should be invalidJSONBytes
in_toto/rulelib.go:1:1: don't use an underscore in package name
in_toto/rulelib_test.go:1:1: don't use an underscore in package name
in_toto/runlib.go:1:1: don't use an underscore in package name
in_toto/runlib.go:25:1: exported function RecordArtifact should have comment or be unexported
in_toto/runlib_test.go:1:1: don't use an underscore in package name
in_toto/util.go:1:1: don't use an underscore in package name
in_toto/util_test.go:1:1: don't use an underscore in package name
in_toto/verifylib.go:6:1: don't use an underscore in package name
in_toto/verifylib.go:522:10: range var authorizedKeyId should be authorizedKeyID
in_toto/verifylib.go:556:21: error strings should not be capitalized or end with punctuation or a newline
in_toto/verifylib.go:571:21: error strings should not be capitalized or end with punctuation or a newline
in_toto/verifylib.go:625:7: range var keyId should be keyID
in_toto/verifylib_test.go:1:1: don't use an underscore in package name
in_toto/verifylib_test.go:551:2: var keyId1 should be keyID1
in_toto/verifylib_test.go:552:2: var keyId2 should be keyID2
in_toto/verifylib_test.go:553:2: var keyId3 should be keyID3
in_toto/verifylib_test.go:619:2: var keyId1 should be keyID1
in_toto/verifylib_test.go:620:2: var keyId2 should be keyID2
in_toto/verifylib_test.go:641:2: var keyId3 should be keyID3
```

**Comment by lukpueh** [comment]
> is the new git history okay or shall I squash more commits? 

looking good. :)

> Will have a look on the coverage next, although I can't explain why the commits have a bad effect on the coverage

Seems like you added new code branches (for error handling) that aren't covered by the tests.

**Comment by shibumi** [comment]
> Seems like you added new code branches (for error handling) that aren't covered by the tests.

Yes, I have a look on them with `go tool cover -html=coverage.out` right now. I am just not sure how I should handle them, because most of the error branches are `os` related (file writing etc).

About the other findings of `golint`. Some of the findings like `inToto` instead of `in_toto` is a breaking change.. not sure if we want to do this. Is software depending on intoto-golang already?

**Comment by shibumi** [comment]
mh interesting, the newest test for an non existing keyfile breaks our appveyor CI :D 
when I run `go test -v ./...` locally it works fine..

**Comment by lukpueh** [comment]
> I am just not sure how I should handle them, because most of the error branches are os related (file writing etc).

We don't need a 100% coverage, if it's too cumbersome to add tests. You can take a look at existing uncovered branches to get a feeling for when we don't write tests. :) 

> About the other findings of golint. Some of the findings like inToto instead of in_toto is a breaking change.. not sure if we want to do this. Is software depending on intoto-golang already?

I think only software that we control, so it should be okay.  But @SantiagoTorres may want to weigh in.

>  it's not between versions, it's between operating systems.. i am playing around with os.NotExistError right now and errors.Unwrap()..

Hm, yes, testing for error messages was a bit quick-and dirty. It would be great if you found a more reliable alternative. Thanks!

**Comment by shibumi** [comment]
> We don't need a 100% coverage, if it's too cumbersome to add tests. You can take a look at existing uncovered branches to get a feeling for when we don't write tests. :)

I am aware of this, I just want to make sure that we have a stable coverage or an increase of coverage, because otherwise I will not get around the CI coverage check. The alternatives are disabling the coverage check or changing the error handling. Instead of returning an error, I could just output a warning, but this seems wrong to me, because we actually want to make sure that the program exits on file permission/file write errors, right?

File: in_toto/canonicaljson.go [hunk 1]
```diff
@@ -93,7 +93,9 @@ func _encodeCanonical(obj interface{}, result *bytes.Buffer) (err error) {
 			// Note: `key` must be a `string` (see `case map[string]interface{}`) and
 			// canonicalization of strings cannot err out (see `case string`), thus
 			// no error handling is needed here.
-			_encodeCanonical(key, result)
+			if err := _encodeCanonical(key, result); err != nil {
```
Comment (shibumi) [inline in_toto/canonicaljson.go:96]: I think I can remove the error handling in this area, according to the comment, right?

File: in_toto/keylib.go [hunk 2]
```diff
@@ -52,10 +52,14 @@ file at path does not exist or is not a PEM formatted RSA public key.
 */
 func (k *Key) LoadPublicKey(path string) error {
 	keyFile, err := os.Open(path)
-	defer keyFile.Close()
 	if err != nil {
 		return err
 	}
+	defer func() {
```
Comment (shibumi) [inline in_toto/keylib.go:58]: My problem right now is in error handling like this. For me it makes sense to at least inform the  user about an error here, but we are unable to properly test error handling for it. At least I don't know how to provoke a keyFile.Close() error

File: .travis.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 language: go
 
 go:
-  - "1.13"
+  - "1.14"
```
Comment (shibumi) [inline .travis.yml:4]: I have increased go version to go 1.14, I hope that's okay.

File: in_toto/keylib.go (hunk 2 above)
Comment (lukpueh) [reply in_toto/keylib.go:58]: That's fine, we really don't have to emulate the error on close. If necessary you can reduce the required coverage in the test suite.

Also, is logging the error the right approach here? Wouldn't it be better to wrap the close error into the error returned by LoadPublicKey?

File: .travis.yml (hunk 3 above)
Comment (lukpueh) [reply .travis.yml:4]: Absolutely! Thanks :)

File: in_toto/keylib.go (hunk 2 above)
Comment (shibumi) [reply in_toto/keylib.go:58]: Mh yes, makes sense. but how would I wrap this into the error returned by the LoadPublicKey() func? Is `defer` passing the error through? I mean If I would do `return err` inside of this anonymous function, would it get passed?

Comment (lukpueh) [reply in_toto/keylib.go:58]: Looks like the `errors` package allows you to do such things. Check out [this article](https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1) and search for *"You can also use this package to wrap multiple errors"*

File: in_toto/canonicaljson.go (hunk 1 above)
Comment (shibumi) [reply in_toto/canonicaljson.go]: removed in newest commit

File: in_toto/keylib.go (hunk 2 above)
Comment (shibumi) [reply in_toto/keylib.go:58]: Ok I can just assign to the err variable from the upper function, like this:
https://play.golang.org/p/BeGOyTH7yPE

I will do this :)

Comment (shibumi) [reply in_toto/keylib.go:58]: Done

**Comment by shibumi** [comment]
@lukpueh can you adjust the code coverage threshold? You can do this as described here:

https://github.com/deepchem/deepchem/issues/648 on https://coveralls.io/

I think a threshold of 0.5% is fine for us

**Comment by lukpueh** [comment]
Thanks for the pointer. I thought we had configured this in .travis.yml. :) I just bumped the threshold to 0.5%

**Comment by shibumi** [comment]
Mh weird.. it still fails for coveralls with -0.2%. Shouldn't the treshold cover this now?

**Comment by lukpueh** [comment]
Can you try again? Apparently coveralls.io's threshold form field didn't like the `%` in my `0.5%` but also didn't care to tell me about it. :/

**Comment by shibumi** [comment]
I will re-open the PR and test..

**Comment by shibumi** [comment]
@lukpueh all tests passed :) ready to merge from my side.

Except you still want this go-lint stuff in the travis build, but I am not sure if we can get *everything* fixed. And it's a binary operator.. either we pass *everything* or it will have no benefit from us.

The last `golint ./...` output brought us this:
```
in_toto/canonicaljson.go:1:1: don't use an underscore in package name
in_toto/canonicaljson.go:85:12: should omit 2nd value from range; this loop is equivalent to `for key := range ...`
in_toto/canonicaljson_test.go:1:1: don't use an underscore in package name
in_toto/examples_test.go:1:1: don't use an underscore in package name
in_toto/hash.go:1:1: don't use an underscore in package name
in_toto/hash.go:21:6: exported type Hash should have comment or be unexported
in_toto/hash.go:77:2: don't use underscores in Go names; var hash_func should be hashFunc
in_toto/hash_test.go:1:1: don't use an underscore in package name
in_toto/hash_test.go:15:2: don't use underscores in Go names; var hash_func should be hashFunc
in_toto/in_toto_test.go:1:1: don't use an underscore in package name
in_toto/keylib.go:1:1: don't use an underscore in package name
in_toto/keylib.go:89:2: var keyIdHashAlgorithms should be keyIDHashAlgorithms
in_toto/keylib_test.go:1:1: don't use an underscore in package name
in_toto/model.go:1:1: don't use an underscore in package name
in_toto/model.go:43:2: struct field KeyId should be KeyID
in_toto/model.go:44:2: struct field KeyIdHashAlgorithms should be KeyIDHashAlgorithms
in_toto/model.go:90:2: struct field KeyId should be KeyID
in_toto/model.go:194:7: exported const LinkNameFormatShort should have comment or be unexported
in_toto/model.go:195:7: exported const SublayoutLinkDirFormat should have comment or be unexported
in_toto/model.go:305:9: range var keyId should be keyID
in_toto/model.go:338:1: comment on exported method Layout.StepsAsInterfaceSlice should be of the form "StepsAsInterfaceSlice ..."
in_toto/model.go:350:1: exported method Layout.InspectAsInterfaceSlice should have comment or be unexported
in_toto/model.go:372:6: range var keyId should be keyID
in_toto/model.go:385:10: if block ends with a return statement, so drop this else and outdent its block
in_toto/model.go:395:10: if block ends with a return statement, so drop this else and outdent its block
in_toto/model.go:427:6: func checkRequiredJsonFields should be checkRequiredJSONFields
in_toto/model_test.go:1:1: don't use an underscore in package name
in_toto/model_test.go:22:2: var invalidJsonBytes should be invalidJSONBytes
in_toto/rulelib.go:1:1: don't use an underscore in package name
in_toto/rulelib_test.go:1:1: don't use an underscore in package name
in_toto/runlib.go:1:1: don't use an underscore in package name
in_toto/runlib.go:5:2: should not use dot imports
in_toto/runlib.go:27:1: exported function RecordArtifact should have comment or be unexported
in_toto/runlib.go:41:2: don't use underscores in Go names; var hash_func should be hashFunc
in_toto/runlib.go:106:1: exported function RecordArtifacts should have comment or be unexported
in_toto/runlib_test.go:1:1: don't use an underscore in package name
in_toto/util.go:1:1: don't use an underscore in package name
in_toto/util_test.go:1:1: don't use an underscore in package name
in_toto/verifylib.go:6:1: don't use an underscore in package name
in_toto/verifylib.go:520:10: range var authorizedKeyId should be authorizedKeyID
in_toto/verifylib.go:554:21: error strings should not be capitalized or end with punctuation or a newline
in_toto/verifylib.go:569:21: error strings should not be capitalized or end with punctuation or a newline
in_toto/verifylib.go:623:7: range var keyId should be keyID
in_toto/verifylib_test.go:1:1: don't use an underscore in package name
in_toto/verifylib_test.go:538:2: var keyId1 should be keyID1
in_toto/verifylib_test.go:539:2: var keyId2 should be keyID2
in_toto/verifylib_test.go:540:2: var keyId3 should be keyID3
in_toto/verifylib_test.go:598:2: var keyId1 should be keyID1
in_toto/verifylib_test.go:599:2: var keyId2 should be keyID2
in_toto/verifylib_test.go:618:2: var keyId3 should be keyID3
```

**Comment by shibumi** [comment]
I would say we fix this go-lint stuff in a separate PR another day and just merge this one here.

**Comment by lukpueh** [comment]
> I would say we fix this go-lint stuff in a separate PR another day and just merge this one here.

Works for me. Thanks for the continued efforts!

**Pull request merged**